### PR TITLE
Improved the API and internal workings

### PR DESCRIPTION
**Pull request by Arcaratus**
-Added IFillable for items that have their own internal LP capacity
-Changed some things that use their own local methods to use the methods within the API
-Added ItemHelper.class which is a class that holds helper methods for many interface methods
-Removed some unnecessary imports
-Switched instances where things were using TileAltar to IBloodAltar
-Cleaned things up here and there
-Added more Javadocs for API methods that didn't already have them

### CODE REVIEW

**Comment by WayofTime** [comment]
I believe changing ItemSigilToggleable to use NBT will change how it needs to be rendered - can you verify that you still have a model when it is active?

File: src/main/java/WayofTime/bloodmagic/api/iface/IFillable.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package WayofTime.bloodmagic.api.iface;
+
+import net.minecraft.item.ItemStack;
+
+/**
+ * Interface used for any item that can store LP in itself
+ */
+public interface IFillable
```
Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/api/iface/IFillable.java]: Seems too generic of a name - needs to describe that it is for LP storage.

File: src/main/java/WayofTime/bloodmagic/api/util/helper/ItemHelper.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import WayofTime.bloodmagic.api.iface.IUpgradeTrainer;
+import WayofTime.bloodmagic.api.livingArmour.LivingArmourHandler;
+import WayofTime.bloodmagic.api.livingArmour.LivingArmourUpgrade;
+import WayofTime.bloodmagic.item.ItemUpgradeTome;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+public class ItemHelper
```
Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/api/util/helper/ItemHelper.java:14]: This class is doing too much, I feel - why should I use the ItemHelper, which is a generic class, when I want to change a specific item (the upgrade tomes)?

File: src/main/java/WayofTime/bloodmagic/item/ItemRitualDiviner.java [hunk 3]
```diff
@@ -206,14 +206,8 @@ public void addInformation(ItemStack stack, EntityPlayer player, List<String> to
...
 
-                int blankRunes = 0;
-                int airRunes = 0;
-                int waterRunes = 0;
-                int fireRunes = 0;
-                int earthRunes = 0;
-                int duskRunes = 0;
-                int dawnRunes = 0;
-                int totalRunes = 0;
+                int blankRunes, airRunes, waterRunes, earthRunes, fireRunes, duskRunes, dawnRunes, totalRunes;
```
Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/item/ItemRitualDiviner.java]: Not the style that's been used so far in the rest of the code.

File: src/main/java/WayofTime/bloodmagic/item/gear/ItemPackSacrifice.java [hunk 4]
```diff
@@ -52,23 +54,10 @@ public ItemPackSacrifice()
             {
                 TileEntity tile = world.getTileEntity(position.getBlockPos());
 
-                if (!(tile instanceof TileAltar))
+                if (tile == null || !(tile instanceof IBloodAltar))
```
Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/item/gear/ItemPackSacrifice.java]: Not needed - instanceof does null checks.

File: src/main/java/WayofTime/bloodmagic/item/gear/ItemPackSelfSacrifice.java [hunk 5]
```diff
@@ -54,29 +64,17 @@ public ItemPackSelfSacrifice()
...
-        } else
+        }
+        else
         {
             if (position.typeOfHit == RayTraceResult.Type.BLOCK)
             {
                 TileEntity tile = world.getTileEntity(position.getBlockPos());
 
-                if (!(tile instanceof TileAltar))
+                if (tile == null || !(tile instanceof IBloodAltar))
```
Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/item/gear/ItemPackSelfSacrifice.java]: Same here

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilToggleableBase.java [hunk 6]
```diff
@@ -45,16 +50,31 @@ public void addInformation(ItemStack stack, EntityPlayer player, List<String> to
...
+        if (!Strings.isNullOrEmpty(getOwnerName(stack)))
             tooltip.add(TextHelper.localizeEffect("tooltip.BloodMagic.currentOwner", PlayerHelper.getUsernameFromStack(stack)));
     }
 
     @Override
-    public List<Pair<Integer, String>> getVariants()
+    @SideOnly(Side.CLIENT)
+    public ItemMeshDefinition getMeshDefinition()
+    {
+        return new CustomMeshDefinitionActivatable("ItemSigil" + WordUtils.capitalize(name));
```
Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilToggleableBase.java:58]: Why are we capitalizing? We usually lowercase everything.

Comment (Arcaratus) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilToggleableBase.java:58]: The Item class names are capitalized in the blockstates

Comment (WayofTime) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilToggleableBase.java:58]: Fair enough

File: src/main/java/WayofTime/bloodmagic/api/impl/ItemSigilToggleable.java [hunk 7]
```diff
@@ -18,30 +20,27 @@
...
-    public void setToggleable()
-    {
-        this.toggleable = true;
     }
 
+    @Override
     public boolean getActivated(ItemStack stack)
     {
-        return stack.getItemDamage() > 0;
+        return stack != null && NBTHelper.checkNBT(stack).getTagCompound().getBoolean(Constants.NBT.ACTIVATED);
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/api/impl/ItemSigilToggleable.java]: `return getActivated(stack)`?

File: src/main/java/WayofTime/bloodmagic/item/ItemBloodOrb.java [hunk 8]
```diff
@@ -65,16 +65,16 @@ public void getSubItems(Item id, CreativeTabs creativeTab, List<ItemStack> list)
             return super.onItemRightClick(stack, world, player, hand);
         }
 
-        if (Strings.isNullOrEmpty(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)))
+        if (Strings.isNullOrEmpty(getOwnerName(stack)))
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBloodOrb.java]: Why was this changed from UUID to name? Same for the two changes below.

File: src/main/java/WayofTime/bloodmagic/item/ItemBoundAxe.java [hunk 9]
```diff
@@ -25,18 +29,11 @@
...
-import WayofTime.bloodmagic.client.mesh.CustomMeshDefinitionActivatable;
-
-import com.google.common.collect.HashMultiset;
-import com.google.common.collect.Multimap;
-import com.google.common.collect.Sets;
 
 import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBoundAxe.java]: Probably shouldn't be doing these import cleanups. Way's settings are different so they'll just keep changing each time.

File: src/main/java/WayofTime/bloodmagic/item/ItemBoundSword.java [hunk 10]
```diff
@@ -97,7 +95,7 @@ public void addInformation(ItemStack stack, EntityPlayer player, List<String> to
 
         tooltip.add(TextHelper.localize("tooltip.BloodMagic." + (getActivated(stack) ? "activated" : "deactivated")));
 
-        if (!Strings.isNullOrEmpty(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)))
+        if (!Strings.isNullOrEmpty(getOwnerName(stack)))
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBoundSword.java]: UUID -> name again

File: src/main/java/WayofTime/bloodmagic/item/ItemBoundTool.java [hunk 11]
```diff
@@ -190,7 +189,7 @@ public void addInformation(ItemStack stack, EntityPlayer player, List<String> to
 
         NBTHelper.checkNBT(stack);
 
-        if (!Strings.isNullOrEmpty(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)))
+        if (!Strings.isNullOrEmpty(getOwnerName(stack)))
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBoundTool.java]: UUID -> name again

File: src/main/java/WayofTime/bloodmagic/item/ItemBoundAxe.java (hunk 9 above)
Comment (Arcaratus) [inline src/main/java/WayofTime/bloodmagic/item/ItemBoundAxe.java]: Didn't do that, IDEA must've, should I change it to java.util.* ?

Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBoundAxe.java]: Look at the changes diff. It's just re-ogranized. This happens in _tons_ of classes.

File: src/main/java/WayofTime/bloodmagic/item/ItemBindableBase.java [hunk 12]
```diff
@@ -29,7 +28,7 @@ public void addInformation(ItemStack stack, EntityPlayer player, List<String> to
     {
         NBTHelper.checkNBT(stack);
 
-        if (!Strings.isNullOrEmpty(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)))
+        if (!Strings.isNullOrEmpty(getOwnerName(stack)))
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBindableBase.java]: UUID -> name

File: src/main/java/WayofTime/bloodmagic/item/ItemBloodOrb.java [hunk 13]
```diff
@@ -65,16 +65,16 @@ public void getSubItems(Item id, CreativeTabs creativeTab, List<ItemStack> list)
...
-        if (Strings.isNullOrEmpty(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)))
+        if (Strings.isNullOrEmpty(getOwnerUUID(stack)))
             return super.onItemRightClick(stack, world, player, hand);
 
         if (world.isRemote)
             return super.onItemRightClick(stack, world, player, hand);
 
-        if (stack.getTagCompound().getString(Constants.NBT.OWNER_UUID).equals(PlayerHelper.getUsernameFromPlayer(player)))
-            NetworkHelper.setMaxOrb(NetworkHelper.getSoulNetwork(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)), getOrbLevel(stack.getItemDamage()));
+        if (PlayerHelper.getUsernameFromStack(stack).equals(PlayerHelper.getUsernameFromPlayer(player)))
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBloodOrb.java]: UUID -> Name (I mentioned both of these last time)

File: src/main/java/WayofTime/bloodmagic/item/ItemBloodOrb.java [hunk 14]
```diff
@@ -65,16 +65,16 @@ public void getSubItems(Item id, CreativeTabs creativeTab, List<ItemStack> list)
...
+            NetworkHelper.setMaxOrb(NetworkHelper.getSoulNetwork(PlayerHelper.getUsernameFromStack(stack)), getOrbLevel(stack.getItemDamage()));
 
-        NetworkHelper.getSoulNetwork(stack.getTagCompound().getString(Constants.NBT.OWNER_UUID)).addLifeEssence(200, getMaxEssence(stack.getItemDamage()));
+        NetworkHelper.getSoulNetwork(PlayerHelper.getUsernameFromStack(stack)).addLifeEssence(200, getMaxEssence(stack.getItemDamage()));
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBloodOrb.java]: UUID -> Name (I mentioned both of these last time)

File: src/main/java/WayofTime/bloodmagic/item/ItemBoundShovel.java [hunk 15]
```diff
@@ -24,18 +28,11 @@
...
-import com.google.common.collect.HashMultiset;
-import com.google.common.collect.Multimap;
-import com.google.common.collect.Sets;
 
 import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
 
 public class ItemBoundShovel extends ItemBoundTool implements IMeshProvider
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/ItemBoundShovel.java:39]: Only change is reorganized imports
